### PR TITLE
4621.dev

### PR DESCRIPTION
**Pull request by 4reeman**
https://redmine.devbranch.work/issues/4621

### CODE REVIEW

**Comment by 4reeman** [comment]
#28

File: web/modules/custom/pulses_weather/src/PulsesCityProvider.php [hunk 1]
```diff
@@ -0,0 +1,148 @@
...
+   *
+   * @throws \Exception
+   */
+  public function getData() {
+    $user = $this->user->id();
+    $exist = $this->dataExist($user);
+    $default_city = $this->configFactory->get(WeatherSettingsForm::SETTINGS)->get('city');
+    $city = $this->getCityName() ?: $default_city;
+    $language = $this->user->getPreferredLangcode();
+    if (!$exist) {
```
Comment (devbranch-vitaliy) [inline web/modules/custom/pulses_weather/src/PulsesCityProvider.php]: We need to get cities only when data doesn't exist.
```suggestion
    if (!$exist) {
      $default_city = $this->configFactory->get(WeatherSettingsForm::SETTINGS)->get('city');
      $city = $this->getCityName() ?: $default_city;
      $language = $this->user->getPreferredLangcode();
    if (!$exist) {
```

**Review by devbranch-vitaliy — CHANGES_REQUESTED** [review]

**Comment by 4reeman** [comment]
@todo create one more column in custom database table for user_ip. It must be implemented because of unauthorized users have the same user_id. Since city for first unauthorized user (user_id = 0) will save in database it won`t change for another unauthorized users because of checking for the existence of an entry in the database of the city of this user is based on user_id.

File: web/modules/custom/pulses_weather/src/PulsesWeatherProvider.php [hunk 2]
```diff
@@ -0,0 +1,80 @@
...
+    $this->dateTime = $date_time;
+  }
+
+  /**
+   * {@inheritDoc}
+   */
+  public function getWeatherInfo($api_key) {
+    if ($cache = $this->cache->get($api_key)) {
+      return $cache->data;
+    }
```
Comment (devbranch-vitaliy) [inline web/modules/custom/pulses_weather/src/PulsesWeatherProvider.php]: Think about cache ID a little bit more ad say me why your implementation is wrong :)

**Review by devbranch-vitaliy — CHANGES_REQUESTED** [review]

**Review by devbranch-vitaliy — APPROVED** [review]

**Pull request merged**